utils/vector.hpp

```diff
@@ -0,0 +1,366 @@
+
+/* Constructors */
+template <class T>
+Vector<T>::Vector()
+    : size(0), capacity(0), array(nullptr)
+{}
+
+
+template<class T>
+Vector<T>::Vector(const Vector<T>& vec)
+    : size(vec.size), capacity(vec.capacity)
+{
+    for(unsigned int i = 0; i < vec.size; i++)
+    {
+        array[i] = vec.array[i];
+    }
+}
+
+
+template<class T>
+Vector<T>::Vector(Vector<T>&& vec)
+    : size(vec.size), capacity(vec.capacity), array(vec.array)
+{
+    vec.array = nullptr;
+}
+
+
+template<class T>
+Vector<T>::Vector(unsigned int size)
+    : size(size), array(nullptr)
+{resize(size);}
+
+
+
+/* Destructor */
+template<class T>
+Vector<T>::~Vector()
+{
+    delete[] array;
+}
+
+
+
+/* Capacity */
+template<class T>
+unsigned int Vector<T>::getSize() const
+{return size;}
+
+
+template<class T>
+unsigned int Vector<T>::getCapacity() const
+{
+    return capacity;
+}
+
+
+template<class T>
+void Vector<T>::resize(unsigned int newSize)
+{
+    if(newSize < size)
+        size = newSize;
+    else if(newSize > size)
+    {
+        //Reallocate the array
+        T* newArray = new T[newSize];
+
+        //Copy the values of the existing array to new array
+        for(unsigned int i = 0; i < size; i++)
+        {
+           newArray[i] = array[i];
+        }
+
+        //Change the data of the vector
+        delete[]   array;
+        array    = newArray;
+        capacity = newSize;
+        size     = newSize;
+    }
+}
+
+
+template<class T>
+void Vector<T>::resize(unsigned int newSize, const T& value)
+{
+    if(size > newSize)
+        size = newSize;
+    else if(size < newSize)
+    {
+        //Reallocate the array
+        T* newArray = new T[newSize];
+
+        //Copy the values of the existing array to new array
+        for(unsigned int i = 0; i < size; i++)
+        {
+            newArray[i] = array[i];
+        }
+
+        //The left over elements will contain the specified value
+        for(unsigned int i = size; i < newSize; i++)
+        {
+            newArray[i] = value;
+        }
+
+        //Change the data of the vector
+        delete[] array;
+        array    = newArray;
+        capacity = newSize;
+        size     = newSize;
+    }
+}
+
+
+template<class T>
+void Vector<T>::reserve(unsigned int newCap)
+{
+    if(capacity < newCap)
+    {
+        //Keep multiplying capacity by 2 until it's larger than the new capacity
+        while(capacity < newCap)
+            capacity *= 2;
+
+        //Reallocate the array
+        T* newArray = new T[capacity];
+
+        //Copy the values of the existing array to new array
+        for(unsigned int i = 0; i < size; i++)
+        {
+            newArray[i] = array[i];
+        }
+
+        //Delete old array and assign new array to vector
+        delete[] array;
+        array = newArray;
+    }
+}
+
+
+
+/* Modifiers */
+template<class T>
+void Vector<T>::push_back(const T& value)
+{
+    if(capacity == size)
+    {
+     resize(size + 1);
+    }
+    array[size] = value;
+}
```

Comment:
This whole function doesn't work. There must be an ``else`` statement incrementing the size (``size++;``).
If you have a capacity of 50 and but no elements (``size = 0``) you will overwrite the same element over and over again.

In addition, since ``size`` is just a number telling how many elements there are in the vector, it cannot be used directly as an index. Therefore, when using it, you must subtract one from it (``array[size - 1] = value;``).